### DISCUSSION TITLE
useQuery time delay of data fetching

### DISCUSSION THREAD

**Discussion by leonelcaroc** [TanStack/query#4930] [General] [answered]
How to delay isLoading state by 5 secs so that the user can see the UI loading component? I'm using useQuery

**Comment by TkDodo** [comment, marked as answer]
That's a ui problem that you have to solve on the ui layer.

**Comment by louis-young** [comment]
As above, deferring a loading state is something you'd need to handle yourself; this isn't the responsibility of TanSack Query.

**Comment by andymerskin** [comment]
@TkDodo / @louis-young  - I disagree, and here's why (hear me out):

This very much feels like an appropriate problem for Query to solve, similar to handling optimistic updates. It's a common problem that's tedious to solve on your own that Query could really shine providing to users.

### The problem
Solving this common problem in the UI code is tedious by itself, because it requires you write a `useEffect` block (React) that watches several pieces of state from Query, including `isLoading`, and `isFetching` — along with any other weird edge cases needing Query's state.

Then, within that block you must setup the timer itself and manage the state in your component manually. After all of that, you've got yourself a pretty hairy mess:
```typescript
const { isFetching, isLoading } = useQuery(...);
const [fetchTime, setFetchTime] = useState(0);
const intervalRef = useRef();

useEffect(() => {
  if (!isFetching || !isLoading) {
    clearInterval(intervalRef.current);
    return;
  }

  intervalRef.current = setInterval(() => {
    setFetchTime(current => current + 100);
  }, 100); // every 100 milliseconds

  return () => clearInterval(intervalRef.current);
}, [isFetching, isLoading]);

if (fetchTime >= 3000 && (isFetching || isLoading)) {
  return <LoadingState />;
}

return <IdealState />;
```
This is kind of gross, and very bug-prone since Query isn't handling this for me at a library level. Now, this could be extracted into a hook that takes a `UseQueryResult`, makes its checks, and returns a `fetchTime` value, but it would be brittle.

### Suggested solution:
I'd love to see usage like this:
```typescript
const { fetchTime, isFetching } = useQuery(...);

// only show loading spinner if it's been loading for a long time, otherwise allow the UI to flash the cache updates...
if (fetchTime >= 3000 && isFetching) {
  return <LoadingState />
}

return <IdealState />
```
where `fetchTime` simply resets internally every time a new fetch happens.

**Comment by wilkinsonj** [comment]
Completely disagree that this is a UI problem. What to do with a specific fetchtime is a UI concern. 

Determining the current fetch time? Is not.

**Comment by silogos** [comment]
I solve this problem use Promise.allSettled() at the queryFn function

Example Promise.allSettled(): https://github.com/silogos/react-tmdb-app/blob/b13955a01d575479f13ab4af1963990e1bcffc2e/src/hooks/useFetch.ts#L34

**Comment by amoretti-dev** [comment]
I solved using `rxjs` for wrapping the fetch fn and using `delay` operator:

```ts
export interface AsyncSearchInputProps<T> {
  fetcher: (query?: string) => Promise<T[]>;
  ...
  resultDelay?: number;
}
export function AsyncSearchInput<T>({
  ...
  fetcher,
  resultDelay = 0
}}: AsyncSearchInputProps<T>) {
  ...
  const fetchFn =
    resultDelay > 0
      ? () =>
          lastValueFrom(
            from(fetcher(debouncedSearchTerm)).pipe(delay(resultDelay))
          )
      : fetcher;
      
  const {
    data: options = [],
    isLoading,
    isError,
    error,
    refetch,
  } = useQuery({
    queryFn: () => fetchFn(debouncedSearchTerm),
    ...
  });
...
}
```
